### PR TITLE
Put assets (modules, logo, PrestaShop versions) in the bucket

### LINKED ISSUE #3: Unless explicitly specified, do not download/generate files that are already downloaded/generated

### PR DESCRIPTION
**Pull request by atomiix**
| Questions         | Answers
| ------------------| -------------------------------------------------------
| Description?      | The main purpose of this PR is to upload module zips and logo & PrestaShop zips to the bucket and make the API returns the url of those files instead of using Github files. It also add some optimizations and add `download_url` & `stability` information on the `/prestashop` & `/prestashop/{stable,rc,beta}` endpoints.
| Type?             | improvement
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #3 #10 & #11 
| How to test?      |  See below
| Possible impacts? | 

### How to test
1. Replace the url in `modules/ps_distributionapiclient/src/DistributionApi.php` line 36 by:
```diff
-private const API_ENDPOINT = 'https://api.prestashop-project.org';
+private const API_ENDPOINT = 'https://integration-api.prestashop-project.org';
```
2. Clear cache: `rm -rf var/cache/*`
3. Fully uninstall any native modules (psgdpr for instance)
4. You should be able to install it back when `ps_distributionapiclient` is enable

### CODE REVIEW

File: src/Command/UploadAssetsCommand.php [hunk 1]
```diff
@@ -14,18 +14,68 @@ class UploadAssetsCommand extends Command
...
+            $output->writeln(sprintf(
+                '<question>Did you run the `%s` command?</question>',
+                DownloadNewPrestaShopReleasesCommand::getDefaultName()
+            ));
+        }
+
+        foreach ($prestashopZips as $prestashopZip) {
+            $filename = self::PRESTASHOP_ASSETS_PREFIX . substr($prestashopZip->getPathname(), strlen($this->prestaShopDir) + 1);
+            $output->writeln(sprintf('<info>Upload file %s</info>', $filename));
+            $this->bucket->upload($prestashopZip->getContents(), ['name' => $filename]);
```
Comment (matks) [inline src/Command/UploadAssetsCommand.php:61]: I am surprised we dont need a try/catch for network issues

File: src/Util/ModuleUtils.php [hunk 2]
```diff
@@ -65,6 +70,30 @@ public function getVersions(string $moduleName, bool $withAssetOnly = true): arr
         ), $versions);
     }
 
+    public function getFromBucket(): ModuleCollection
+    {
+        $modules = $this->bucket->objects(['prefix' => 'assets/modules/']);
+        $list = [];
+        foreach ($modules as $module) {
+            $parts = explode('/', $module->info()['name']);
+            if (!isset($list[$parts[count($parts) - 3]])) {
```
Comment (matks) [inline src/Util/ModuleUtils.php:79]: Why 3 🤔

File: src/Util/ModuleUtils.php [hunk 3]
```diff
@@ -65,6 +70,30 @@ public function getVersions(string $moduleName, bool $withAssetOnly = true): arr
...
+                $list[$parts[count($parts) - 3]] = [];
+            }
+            $list[$parts[count($parts) - 3]][] = $parts[count($parts) - 2];
+        }
+
+        $tet = [];
```
Comment (matks) [inline src/Util/ModuleUtils.php:85]: Tet ? 😄

File: src/Util/ModuleUtils.php [hunk 4]
```diff
@@ -74,18 +103,53 @@ public function downloadMainClass(string $moduleName, Version $version): void
...
 
-    /**
-     * @return Module[]
-     */
-    public function getLocalModules(): array
+    public function download(string $moduleName, Version $version): void
+    {
+        $path = join('/', [$this->moduleDir, $moduleName, $version->getTag()]);
+        if (!is_dir($path)) {
+            mkdir($path, 0777, true);
```
Comment (matks) [inline src/Util/ModuleUtils.php:112]: 0777 😄 ?

File: src/Util/ModuleUtils.php [hunk 5]
```diff
@@ -74,18 +103,53 @@ public function downloadMainClass(string $moduleName, Version $version): void
...
+        }
+
+        if ($version->getGithubUrl() === null) {
+            throw new RuntimeException(sprintf('Unable to download %s %s because it has no Github url', $moduleName, $version->getTag()));
+        }
+
+        $response = $this->client->get($version->getGithubUrl());
+        file_put_contents($path . '/' . $moduleName . '.zip', $response->getBody());
```
Comment (matks) [inline src/Util/ModuleUtils.php:120]: You need to check the path is valid before I think

File: src/Util/ModuleUtils.php [hunk 6]
```diff
@@ -74,18 +103,53 @@ public function downloadMainClass(string $moduleName, Version $version): void
...
+        if (!file_exists($path)) {
+            return;
+        }
+
+        $moduleZip = new ZipArchive();
+        $moduleZip->open($path);
+        $icon = $moduleZip->getFromName($moduleName . '/logo.png');
+        $moduleZip->close();
+
+        file_put_contents(join('/', [$this->moduleDir, $moduleName, $version->getTag(), 'logo.png']), $icon);
```
Comment (matks) [inline src/Util/ModuleUtils.php:135]: Same here, check the path

Be paranoid 😄

File: src/Util/PrestaShopUtils.php [hunk 7]
```diff
@@ -71,16 +122,30 @@ public function getLocalVersions(): array
...
     }
 
-    public function setVersionsCompat(PrestaShop $prestaShop): void
+    private function setVersionsCompat(PrestaShop $prestaShop): void
     {
         $parser = (new ParserFactory())->create(ParserFactory::PREFER_PHP7);
-        if (!$content = file_get_contents($this->prestaShopDir . '/' . $prestaShop->getVersion() . '/install_version.php')) {
+        $installZip = new ZipArchive();
+        $installZip->open($this->prestaShopDir . '/' . $prestaShop->getVersion() . '/prestashop.zip');
+        $installZip->extractTo($this->prestaShopDir . '/' . $prestaShop->getVersion() . '/prestashop/');
```
Comment (matks) [inline src/Util/PrestaShopUtils.php:139]: I think we might need try/catch to handle when the archive process fails

**Review by matks — APPROVED** [review]
I approve the PR but

- I added some suggestions to strengthen the code against network failures, bad filesystem paths and process failures (like building a ZIP archive)
- With the addition of the Google Bucket management, is it still easy to work on localhost? I'm worried we're coupling hard the codebase with Google Bucket dependency. How could we work-around that?

File: src/Util/ModuleUtils.php (hunk 4 above)
Comment (atomiix) [reply src/Util/ModuleUtils.php:112]: mmh, yes? What's the issue with this? 😅

File: src/Util/ModuleUtils.php (hunk 3 above)
Comment (atomiix) [reply src/Util/ModuleUtils.php:85]: That's a temporary name that successfully persisted 😆

File: src/Util/ModuleUtils.php (hunk 2 above)
Comment (atomiix) [reply src/Util/ModuleUtils.php:79]: I'll update this method so it will be clearer

File: src/Util/ModuleUtils.php (hunk 5 above)
Comment (atomiix) [reply src/Util/ModuleUtils.php:120]: The path is created before if it doesn't exists

File: src/Util/ModuleUtils.php (hunk 6 above)
Comment (atomiix) [reply src/Util/ModuleUtils.php:135]: The function returns if the path doesn't exists

File: src/Util/PrestaShopUtils.php (hunk 7 above)
Comment (atomiix) [reply src/Util/PrestaShopUtils.php:139]: I'm not sure we want the script to continue if this fails actually 🤔

**Comment by atomiix** [comment]
>I approve the PR but
>
>- I added some suggestions to strengthen the code against network failures, bad filesystem paths and process failures (like building a ZIP archive)
>- With the addition of the Google Bucket management, is it still easy to work on localhost? I'm worried we're coupling hard the codebase with Google Bucket dependency. How could we work-around that?

For the first point, I think we want the script to fail and stop if there is any error so we don't have half a release available

For the second point, yes it's getting more coupled but you can easily create a personal bucket to run it

File: src/Util/ModuleUtils.php (hunk 4 above)
Comment (matks) [reply src/Util/ModuleUtils.php:122]: 0755? 0644? But 0777 is usually NOT the good option

File: src/Util/PrestaShopUtils.php (hunk 7 above)
Comment (matks) [reply src/Util/PrestaShopUtils.php:139]: Even if we want the script to stop on such failure, then I would prefer we make it explicit:
```
try {
} catch (CannotCreateZipArchiveException $exception) {
    throw new Exception('App cannot Create a ZipArchive, please debug');
}
```

**Comment by matks** [comment]
> For the first point, I think we want the script to fail and stop if there is any error so we don't have half a release available

OK but in the logs I wish we could see "Throw Exception at line ..." rather than "Fatal error ..." (see above comment)

**Comment by matks** [comment]
@atomiix If you agree with my suggestion, do we go for QA and you improve error logs later or do you wish to include logs error improvement in another PR? I approved the PR to help you move forward as my feedback is not blocking

**Comment by atomiix** [comment]
We need https://github.com/PrestaShopCorp/terraform-core-oss-distribution-api/pull/7 to be merged to be able to test this PR but yeah, as soon as it's merged, let's go for QA and I'll improve the error logs in another PR ;)

**Review by matks — APPROVED** [review]
@atomiix I approve the PR as it is because the quality is good enough

but I hope that we'll be able to bring improvements it later ^^

What's the next step?

**Comment by atomiix** [comment]
@matks I updated the how to test section, I add the `waiting for QA` label ;)

**Pull request merged**